core/clients/key_flow.go

```diff
@@ -0,0 +1,359 @@
+		}
+	}()
+	return c.parseTokenResponse(res)
+}
+
+// generateSelfSignedJWT generates JWT token
+func (c *KeyFlow) generateSelfSignedJWT() (string, error) {
+	claims := jwt.MapClaims{
+		"iss": c.key.Credentials.Iss,
+		"sub": c.key.Credentials.Sub,
+		"jti": uuid.New(),
+		"aud": c.key.Credentials.Aud,
+		"iat": jwt.NewNumericDate(time.Now()),
+		"exp": jwt.NewNumericDate(time.Now().Add(10 * time.Minute)),
+	}
+	token := jwt.NewWithClaims(jwt.SigningMethodRS512, claims)
+	token.Header["kid"] = c.key.Credentials.Kid
+	tokenString, err := token.SignedString(c.privateKey)
+	if err != nil {
+		return "", err
+	}
+	return tokenString, nil
+}
+
+// requestToken makes a request to the SA token API
+func (c *KeyFlow) requestToken(grant, assertion string) (*http.Response, error) {
+	body := url.Values{}
+	body.Set("grant_type", grant)
+	body.Set("assertion", assertion)
+	payload := strings.NewReader(body.Encode())
+	req, err := http.NewRequest(http.MethodPost, c.config.TokenUrl, payload)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	return c.doer(&http.Client{}, req, c.config.ClientRetry)
+}
+
+// parseTokenResponse parses the response from the server
+func (c *KeyFlow) parseTokenResponse(res *http.Response) error {
+	if res == nil {
+		return fmt.Errorf("received bad response from API")
+	}
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("received: %+v", res)
+	}
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+	c.token = &TokenResponseBody{}
+	return json.Unmarshal(body, c.token)
+}
+
+// validateToken returns true if token is valid
+func (c *KeyFlow) validateToken(token string) (bool, error) {
+	if token == "" {
+		return false, nil
+	}
+	if _, err := c.parseToken(token); err != nil {
+		if strings.Contains(err.Error(), "401") {
+			c.token = new(TokenResponseBody)
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// parseToken parses and validates a JWT token
+func (c *KeyFlow) parseToken(token string) (*jwt.Token, error) {
+	b, err := c.getJwksJSON(token)
+	if err != nil {
+		return nil, err
+	}
+	var jwksBytes = json.RawMessage(b)
+	jwks, err := keyfunc.NewJSON(jwksBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return jwt.Parse(token, jwks.Keyfunc)
+}
+
+func (c *KeyFlow) getJwksJSON(token string) (jwks []byte, err error) {
+	req, err := http.NewRequest("GET", c.config.JWKSUrl, http.NoBody)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+	res, err := c.doer(&http.Client{}, req, c.config.ClientRetry)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		tempErr := res.Body.Close()
+		if tempErr != nil {
+			jwks = nil
+			err = fmt.Errorf("closing get jwks response: %w", tempErr)
+		}
+	}()
+	if res.StatusCode == 200 {
+		return io.ReadAll(res.Body)
+	}
+	return nil, fmt.Errorf("error: %s", res.Status)
```

Comment:
I'd remove the `error:` part, if you've got nothing to add, just write the status

Also, to be more Go-ish, swap the if condition to be `if res.StatusCode != 200`